✨ NEW: Introducing jupyterbook-latex and added handling of tableofcontents for latex

AakashGfude opened the pull request:
This PR brings a range of improvements to `PDF` builds via `LaTeX` by incorporating the [jupyterbook-latex](https://github.com/executablebooks/jupyterbook-latex) extension. This initial release is focused on bringing harmony to `html` and `latex` outputs (through support for `part`/`chapter` structures in the `_toc.yml`), support for the jupyter-book directive `tableofcontents`, and support for commonly used tags on code-cell directives such as `hide-cell`. 

This initial release targets `whole of project` pdf files only. 

1. adds the [jupyterbook-latex](https://github.com/executablebooks/jupyterbook-latex) extension which provides improvements for `pdf` and `latex` generation. Activation of this package is configurable with the `use_jupyterbook_latex` key (turned on by default). Further information for `jupyterbook-latex` can be read here: https://github.com/executablebooks/jupyterbook-latex/blob/master/docs/intro.md. This package adds support for document structures that are [recommended for jupyter-book](https://jupyterbook.org/customize/toc.html)
2. adds support for the `tableofcontents` directive for LaTeX by creating a list of reference nodes such as. 
<img width="721" alt="Screen Shot 2021-02-02 at 10 56 54 am" src="https://user-images.githubusercontent.com/6542997/106533008-6bc2a800-6545-11eb-93cc-13fc542fb037.png">

We have documented some [design decisions and future work in this discussion thread](https://github.com/executablebooks/meta/discussions/270). 

TODO:

- [x] only include entries that are children of the document in which directive is used
- [x] Add tests 

fixes https://github.com/executablebooks/jupyter-book/issues/847
fixes https://github.com/executablebooks/jupyter-book/issues/810
fixes https://github.com/executablebooks/jupyter-book/issues/1192

AakashGfude commented:
Have also introduced `jupyterbook_latex` here as well, the reason for which is outlined in https://github.com/executablebooks/jupyter-book/pull/1125 . 

The styling of tableofcontents directive follows the default styling of sphinx for now. (the title 'Table of Contents' is part of markdown here not the directive)
<img width="801" alt="Screen Shot 2021-02-01 at 1 23 36 pm" src="https://user-images.githubusercontent.com/6542997/106407637-b2a29600-6490-11eb-8fa6-93a519ebeee1.png">

cc : @mmcky @choldgraf @chrisjsewell

mmcky commented:
@AakashGfude this is looking good. May I suggest you edit the top (main) comment field with a full description including the inclusion of `jupyterbook-latex`? That will make it easier for putting together releases as this is a big improvement for `pdf` outputs. Will this also include `sphinx-multitoc-numbering` and change default numbering behaviour? Let's add a section for any proposed changes to defaults etc. 

This will supersede #1125 right?

AakashGfude commented:
> @AakashGfude this is looking good. May I suggest you edit the top (main) comment field with a full description including the inclusion of `jupyterbook-latex`? That will make it easier for putting together releases as this is a big improvement for `pdf` outputs. Will this also include `sphinx-multitoc-numbering` and change default numbering behaviour? Let's add a section for any proposed changes to defaults etc.
> 
> This will supersede #1125 right?

Thanks @mmcky , have updated the top comment. No, I have not introduced `spinx-multitoc-numbering` with this. Will create a new PR for that. 
Yes, it will supersede https://github.com/executablebooks/jupyter-book/pull/1125 . 

I have included the doc page, where all the features are outlined. Hopefully, that can outline the proposed changes?
We can also edit the docs if some things are not clear.

chrisjsewell commented on jupyter_book/directive/toc.py:
```diff
@@ -25,26 +31,136 @@ def run(self):
+                    return filtered_toc
+        return
+
+    def _has_toc_yaml(self, subnode, tocdict, depth):
```
I think this method, together with `_handle_toc_header`, should be a standalone function. Then you can you add some unit tests, e.g. constructing example globaltoc dicts and testing the output `subnode.pformat()`. Because at the moment there is only a few integration tests where it is difficult to test a large range of possibilities and to understand what the functions do.

Also can you try "typing" your methods and providing docstring explanations of the inputs, e.g. something like

```python
from typing import Dict

from docutils import nodes

def _has_toc_yaml(self, subnode: nodes.Element, tocdict: Dict[str, nodes.Element], depth: int) -> None:
   """constructs toc nodes from globaltoc dict

	:param subnode: this is....
    :param tocdict: this is....
    :param depth: this is....
   """
``` 

You should be trying to type all your code and using mypy, for example markdown-it-py and myst-parser are now fully typed (https://github.com/executablebooks/MyST-Parser/commit/07ac8596f2171be993c931b4e58de93ed940c9c0) and this will eventually be the same for myst-nb

AakashGfude replied:
Great. Thanks, @chrisjsewell . On it

AakashGfude replied:
Hey @chrisjsewell, have typed the functions and added tests for individual ones. let me know if it is looking better.

chrisjsewell reviewed (changes requested):
Looking good cheers, just a request for documenting/typing the code a bit more, and some more "lower level" unit tests

chrisjsewell commented:
Also in https://github.com/executablebooks/jupyterbook-latex/blob/master/docs/intro.md it says "jupyterbook-latex is in an active development stage and may change rapidly.", and it says about "git cloning" to install.

a) With these changes, is this package more stable now, and so we don't have to essentially warn people away from using anymore
b) Is there a way we can have an example build that be can be linked to from the documentation, prefereably of the actual jupyter-book documentation, so that people can see for themselves what the output looks like
c) given jupyter-latex is now part of the required installs, we can probably remove the installation instructions
d) also given it is part of the required installs, we need it to have a conda-forge package. Is this the case?

mmcky commented:
@AakashGfude is `TexSoup` a dependency that needs to be added?

AakashGfude commented:
> @AakashGfude is `TexSoup` a dependency that needs to be added?

@mmcky have added the dependency here https://github.com/executablebooks/jupyter-book/pull/1167/files#diff-60f61ab7a8d1910d86d9fda2261620314edcae5894d5aaa236b821c7256badd7R20 , only for the testing case.

chrisjsewell commented on jupyter_book/directive/toc.py:
```diff
@@ -22,29 +29,154 @@ def run(self):
+
+
 class SwapTableOfContents(SphinxTransform):
     default_priority = 700
```
Can you add a comment on why 700 was chosen. Is there anything specific it needs to come before/after?

AakashGfude replied:
No particular reason. I think I was trying something out but the priority did not matter, should probably remove it?

chrisjsewell replied:
I think its required to have a priority, you see the range categories here https://www.sphinx-doc.org/en/master/extdev/appapi.html?highlight=config-inited#sphinx.application.Sphinx.add_transform

AakashGfude replied:
@chrisjsewell can you help with what priority to choose 😬. Ideally would want it to be the lowest possible one for the given functionality, right?

mmcky replied:
thanks @chrisjsewell that is a helpful link. @AakashGfude perhaps you should set this in the range of `900-999` unless you don't need `referencing` resolution given `LaTeX` probably doesn't need that.

mmcky commented:
> > @AakashGfude is `TexSoup` a dependency that needs to be added?
> 
> @mmcky have added the dependency here https://github.com/executablebooks/jupyter-book/pull/1167/files#diff-60f61ab7a8d1910d86d9fda2261620314edcae5894d5aaa236b821c7256badd7R20 , only for the testing case.

Hmm ... I just had a failure when using `tox` due to `TexSoup` missing.

chrisjsewell commented:
> Hmm ... I just had a failure when using tox due to TexSoup missing.

did you run `tox -r` to update the environment?

mmcky commented:
@AakashGfude doing some testing on [lecture-python.myst](https://github.com/QuantEcon/lecture-python.myst) I am finding:

1. `unicode` errors issued by latex (I know you're looking into this - as per Slack discussion)
2. `execution` via `jupyter-cache` seems to run fresh each time. 

I'll continue to dig a little into these issues

AakashGfude commented:
> @AakashGfude doing some testing on [lecture-python.myst](https://github.com/QuantEcon/lecture-python.myst) I am finding:
> 
> 1. `unicode` errors issued by latex (I know you're looking into this - as per Slack discussion)

oopsie. sorry had not setup `jupyterbook-latex` as an extension in `__init__` of this PR. have done it now. https://github.com/executablebooks/jupyter-book/pull/1167/commits/61f4cb9a99222fdd299e1e30cb4b98c6b9e4b89a
So, should use the extension now  by default.

> 2. `execution` via `jupyter-cache` seems to run fresh each time.
Since your build was not using `jupyterbook-latex` earlier, i suspect this issue is not related to this PR then.

mmcky commented:
thanks @AakashGfude building nicely now after `git pull` -- I will review the `pdf` once it is built. Also the execution is no longer running each time 🥳

AakashGfude commented:
> Also in https://github.com/executablebooks/jupyterbook-latex/blob/master/docs/intro.md it says "jupyterbook-latex is in an active development stage and may change rapidly.", and it says about "git cloning" to install.
> a) With these changes, is this package more stable now, and so we don't have to essentially warn people away from using anymore

I belive with our testing in different projects its stable enough.  So, I reckon we can remove this.

> b) Is there a way we can have an example build that be can be linked to from the documentation, preferably of the actual jupyter-book documentation, so that people can see for themselves what the output looks like.

Should we add a button to the theme, to download the whole repo as pdf using this extension as well? 

> c) given jupyter-latex is now part of the required installs, we can probably remove the installation instructions.

have done these changes in this PR https://github.com/executablebooks/jupyterbook-latex/pull/37/files#diff-22760d417a52c66f14bee182587d458d0737a616dd14cb09748b4c725fc5809fR12 .

mmcky commented:
> Should we add a button to the theme, to download the whole repo as pdf using this extension as well?

@AakashGfude perhaps there should be an `option` that users can switch on for `sphinx-book-theme` to be powered either by `javascript` or by `latex` for pdf support. However as `jupyterbook-latex` currently doesn't support individual `pdf` files yet -- we should wait on this feature so it is a seamless switch. I think it will be OK for now for users to fetch the `pdf` using `jb build ./ --builder=pdflatex` and going to `_build/latex`. What do you think?

mmcky commented:
@AakashGfude is there a way to switch this feature off just in case a user has configured something that they are using with the existing setup? I think this should be switched `on` by default to support `files` and `part/chapter` structures in the `_toc.yml` -- but it might be nice to be able to switch it off (if needed).

AakashGfude commented:
> @AakashGfude is there a way to switch this feature off just in case a user has configured something that they are using with the existing setup? I think this should be switched `on` by default to support `files` and `part/chapter` structures in the `_toc.yml` -- but it might be nice to be able to switch it off (if needed).

have added a key called `use_jupyterbook_latex` which is turned on by default.

mmcky commented on tests/test_tocdirective/test_toc_urllink.html:
```diff
@@ -1,6 +1,6 @@
 <div class="tableofcontents-wrapper compound">
 <ul>
-<li class="toctree-l1"><a class="reference internal" href="content1.html">Content1</a></li>
+<li class="toctree-l1"><a class="reference internal" href="content1.html">Content 1</a></li>
```
@AakashGfude is it expected that `Content 1` here has a space but none of the others contain a space.

AakashGfude replied:
@mmcky yes. Because the `_toc.yml` for this file, does not have a `-title` specified for `Content1`, so it takes the title from the doc itself, which is `Content 1`. Did this change in `toc`  to increase coverage of the code.

mmcky replied:
ok - great. Thanks @AakashGfude

mmcky commented on .github/workflows/pdf.yml:
```diff
@@ -78,3 +78,6 @@ jobs:
       with:
         name: PDF_LATEX
         path: docs/_build/latex/book.pdf
+    - uses: actions/checkout@v2
```
Checkout is for setting up the `repository` so what you're trying to do here won't work. I would delete this addition and keep the `artifact` upload above. 

The issue here is that the docs are probably built through `readthedocs` so we will need to check how execution is supported there and if `xelatex` is available for building.

mmcky replied:
We should see if `artifacts` are linkable perhaps -- otherwise we may need to save that `artifact` through the `github action` to some linkable location.

mmcky replied:
It looks like the `link` are dependent on `run` numbers so are probably `dynamic`

https://github.com/executablebooks/jupyter-book/suites/2074361830/artifacts/41968814

This [issue](https://github.com/actions/upload-artifact/issues/50) kind of helps but hard to add to the `myst` file. But perhaps we should research a way to get access to these artifacts generated on the `master` branch to get a stable `html link`

mmcky replied:
@choldgraf would you be able to provide us with access to `rtd`?

AakashGfude replied:
@mmcky I was just trying to initiate a discussion on this with that commit. But I reckon we can also do it in a new PR (after this is merged), to prevent this one from having too many lateral features?

mmcky replied:
sure @AakashGfude we can add this to the docs in a new PR if you like. Then we can freeze this for final review and work out the `rtd` context separately.

mmcky commented:
hey @AakashGfude is this ready for another round of reviews?

AakashGfude commented:
> hey @AakashGfude is this ready for another round of reviews?

@mmcky yes it is.

mmcky commented on tests/books/toc/_toc_urllink.yml:
```diff
@@ -14,5 +13,5 @@ sections:
   - file: subfolder/asubpage
     title: Asubpage
-  - url: https://jupyterbook.org/file-types/markdown.html
+  - url: www.jupyterbook.org/file-types/markdown.html
```
@AakashGfude what is driving this change?

AakashGfude replied:
@mmcky just to test out a faulty URL. 
In case of HTML, if a url does not have `http` or `https` , it is not included in the toc. The same has been done for `tableofocontents` directive. So, the above change was testing out this case.

mmcky reviewed (approved):
@AakashGfude this is looking great. I have reviewed this PR now against three real world projects and it is producing nice `pdf` output in a structure that mimics the `html` version.
I think this is good to 🚢 - and we can continue to make improvements via PR. 

- [ ] update `issues` that are solved with useful update comment and link to this PR. 

Thanks for working on this @AakashGfude

mmcky reviewed (commented):
@AakashGfude we should also check:

- [x] https://github.com/executablebooks/jupyter-book/issues/1192 

before merging is working as an excellent test case.

AakashGfude commented:
> * [ ]  Also do we need to pin the version of `jupyterbook-latex`?

I think the best would be to use the latest version. no version pinning required.

mmcky commented:
@AakashGfude I think version pinning is a good idea -- while it increases maintenance work it improves stability and makes bug tracking a lot easier when all dependencies are version tracked.

AakashGfude commented:
> @AakashGfude I think version pinning is a good idea -- while it increases maintenance work it improves stability and makes bug tracking a lot easier when all dependencies are version tracked.

hmmm. This makes sense.
 
My reasoning was that since the release of versions is quicker in this repo as it is a comparatively smaller one, we should not pin the version for now. Also since, people might encounter some issues initially when this is merged, we might need to make releases quicker. And if we want those bug fixes to be available, we will have to increment the version in `jupyter-book` and wait for a version of `jupyter-book` to be released. Which might slow the process? Am i making sense?

chrisjsewell commented:
> Maybe we wait for a few weeks and see if it's working stable enough for the community, and then pin a version?

no we should always pin, if you are only making bug fixes the they will be patch version so you wouldn't need to update jupyter-book, i.e. if you use then `jupyterbook-latex~=0.1.0` then you can update `jupyterbook-latex` with a bug fix to `0.1.1` and that would be fine, but then if you are making a back compatible change you update to `0.2.0` and update jupter-book as appropriate

chrisjsewell commented:
while I'm here have you addressed all the points in https://github.com/executablebooks/jupyter-book/pull/1167#issuecomment-776339722 yet 😬?

AakashGfude commented:
> > Maybe we wait for a few weeks and see if it's working stable enough for the community, and then pin a version?
> 
> no we should always pin, if you are only making bug fixes the they will be patch version so you wouldn't need to update jupyter-book, i.e. if you use then `jupyterbook-latex~=0.1.0` then you can update `jupyterbook-latex` with a bug fix to `0.1.1` and that would be fine, but then if you are making a back compatible change you update to `0.2.0` and update jupter-book as appropriate

Great. Thank you for further clarification :)

AakashGfude commented:
> > no we should always pin, if you are only making bug fixes the they will be patch version so you wouldn't need to update jupyter-book, i.e. if you use then `jupyterbook-latex~=0.1.0` then you can update `jupyterbook-latex` with a bug fix to `0.1.1` and that would be fine, but then if you are making a back compatible change you update to `0.2.0` and update jupter-book as appropriate

The present version of `jupyterbook-latex` is 0.1.8 . I have to make a new release tomorrow with some additional fixes. I will upgrade it to 0.2.0 , so further minor fixes will not need any updates to `jupyter-book`. Sounds good?

AakashGfude commented:
> while I'm here have you addressed all the points in [#1167 (comment)](https://github.com/executablebooks/jupyter-book/pull/1167#issuecomment-776339722) yet 😬?

I have for the most part. The discussion about those points start from this comment I think https://github.com/executablebooks/jupyter-book/pull/1167#issuecomment-777944537 . 
Have not done the following point 😬:
> is there a way we can have an example build that be can be linked to from the documentation, prefereably of the actual jupyter-book documentation, so that people can see for themselves what the output looks like`

As we were thinking of doing it in a separate PR.

mmcky commented:
> > > no we should always pin, if you are only making bug fixes the they will be patch version so you wouldn't need to update jupyter-book, i.e. if you use then `jupyterbook-latex~=0.1.0` then you can update `jupyterbook-latex` with a bug fix to `0.1.1` and that would be fine, but then if you are making a back compatible change you update to `0.2.0` and update jupter-book as appropriate
> 
> The present version of `jupyterbook-latex` is 0.1.8 . I have to make a new release tomorrow with some additional fixes. I will upgrade it to 0.2.0 , so further minor fixes will not need any updates to `jupyter-book`. Sounds good?

For this to work out for automatic minor I think you'll need to pin using the pattern `v0.2.*`

mmcky commented:
> > while I'm here have you addressed all the points in [#1167 (comment)
> Have not done the following point 😬:
> 
> > is there a way we can have an example build that be can be linked to from the documentation, prefereably of the actual jupyter-book documentation, so that people can see for themselves what the output looks like`
> 
> As we were thinking of doing it in a separate PR.

@chrisjsewell @AakashGfude and I reviewed this. The `pdf` is being built in tests - but we need a location that we can link too for the `docs`. We were trying to use `github artifacts` using the `upload-artifacts` action but unfortunately the `url` is linked to the run number so is always changing. Also we don't have access to `RTD` to check if we have control over the build there.

AakashGfude commented:
@chrisjsewell @mmcky had pinned the version to `~=0.2.0` here. Let me know if there is anything else we need to do here.

mmcky commented:
thanks @AakashGfude for continuing to make improvements as we reviewed issues etc. This is now looking pretty good and is ready for any final review from the team. 

I have documented the [design strategy and next steps for LaTeX improvements](https://github.com/executablebooks/meta/discussions/270) to help give some context and let you know where we are heading.

mmcky commented:
thanks @AakashGfude for adding https://github.com/executablebooks/jupyterbook-latex/pull/49. This looks good to 🚢

choldgraf commented on jupyter_book/__init__.py:
```diff
@@ -40,6 +44,7 @@ def setup(app):
     # Transforms
     app.add_post_transform(SwapTableOfContents)
 
```
This comment seems a bit out of place since there's nothing underneath it....

AakashGfude replied:
you mean `# Transforms`? Should we change it to `#Post transforms`?  I think we were using this comment format to list different kinds of additions to the app.

choldgraf replied:
no just underneath this line there is `# Extensions` and then it goes right to the "return" statment

choldgraf commented on jupyter_book/commands/__init__.py:
```diff
@@ -242,6 +242,8 @@ def build(
         # Builder-specific overrides
         if builder == "pdfhtml":
             config_overrides["html_theme_options"] = {"single_page": True}
+        if builder == "pdflatex":
```
Is this just temporarily because of the sphinxcontrib-bibtex bug? or not?

AakashGfude replied:
Thanks @choldgraf for pointing this out. I will remove this one, as we have already merged a temporary fix for sphinxcontrib-bibtex

choldgraf commented on jupyter_book/config.py:
```diff
@@ -159,6 +159,12 @@ def get_final_config(
     }
 
+    if (
+        "use_jupyterbook_latex" in sphinx_config
```
won't `use_jupyterbook_latex` always be in `sphinx_config` since it's a default? And in that case we could probably just do `sphinx_config.get("use_jupyterbook_latex")` no?

AakashGfude replied:
yup. you are right. have made the changes.

choldgraf commented on jupyter_book/directive/toc.py:
```diff
@@ -22,29 +29,154 @@ def run(self):
 
-class SwapTableOfContents(SphinxTransform):
-    default_priority = 700
+def process_toc_dict(
```
Is this for use with `tableofcontents` or something? Does the apply to non-latex usecases or is it just for latex?

AakashGfude replied:
It is just for latex use case at the moment. The `SwapTableofContents` post-transform handles latex and non-latex differently. This function is used only in latex case. https://github.com/executablebooks/jupyter-book/pull/1167/files#diff-d6d834bc60c86a22e11aed333d22d8ac5f9eb097f8da28d3c299da8e2806adf5R137

choldgraf reviewed (commented):
Hey all - in general this looks good to me! It's a bit unclear what all the `tableofcontents` changes are affecting (e.g. if it's just for latex or also for HTML). I added a few comments in there!

AakashGfude commented:
> Hey all - in general this looks good to me! It's a bit unclear what all the `tableofcontents` changes are affecting (e.g. if it's just for latex or also for HTML). 

It should just be affecting the Latex output, the one which was written earlier for HTML has just been copied in this else case 
https://github.com/executablebooks/jupyter-book/pull/1167/files#diff-d6d834bc60c86a22e11aed333d22d8ac5f9eb097f8da28d3c299da8e2806adf5R158

mmcky commented:
@chrisjsewell @choldgraf I am planning to merge this on `Friday 12th March` unless there are any further comments or objections. Thanks!

choldgraf commented:
+1 from me 👍

mmcky commented:
thanks @AakashGfude for this pull request and enhancements -- and iterating on comments and design. This will be a useful addition and big improvement to LaTeX builds.

mmcky commented:
@chrisjsewell I am not sure why GitHub still had your `change request` as open as @AakashGfude had implemented some changes around your comments. So I dismissed that review (as `stale` to get a green Squash and Merge Status)

chrisjsewell commented:
Heya @AakashGfude @mmcky, yeh in general all looks great thanks 👍 
Two nitpicks though 😬:

>  also given it is part of the required installs, we need it to have a conda-forge package

I don't see a jupyter-latex package on conda yet? Can you address this as a priority

https://github.com/executablebooks/jupyter-book/commit/a3065b134c6f6b1b6b467ca0652778709d156c7f; can you make sure that commit messages adhere to https://github.com/executablebooks/.github/blob/master/CONTRIBUTING.md#commit-messages

mmcky commented:
oh man that commit message is awful. Sorry @chrisjsewell. I should have adjusted the first cell of the PR. Is there a way to amend this?

mmcky commented:
```
I don't see a jupyter-latex package on conda yet?
```

- [ ] @AakashGfude can you organise this as a priority.

chrisjsewell commented:
> can you make sure that commit messages adhere to executablebooks/.github@master/CONTRIBUTING.md#commit-messages

Same goes for @choldgraf with https://github.com/executablebooks/jupyter-book/commit/5de7ca62541ed4641fb6968c5b58830fec2b5c74 😉 it just helps for understanding changes in retrospect and constructing the changelog for a release

chrisjsewell commented:
> I should have adjusted the first cell of the PR. Is there a way to amend this?

you can edit any of the message cells (top-right), but also you can just edit the message directly, before pressing "Confirm squash and merge"

chrisjsewell commented:
> @AakashGfude can you organise this as a priority.

let me know if you need any help, its pretty straightforward

mmcky commented:
> > I should have adjusted the first cell of the PR. Is there a way to amend this?
> 
> you can edit any of the message cells (top-right), but also you can just edit the message directly, before pressing "Confirm squash and merge"

Sorry I meant is there a way to amend the commit message? I'll have a look over the first cell of the PR history too.

chrisjsewell commented:
you can do an interactive rebase of the master branch, although thats a bit naughty 😬 (I'm not sure if just changing the message changes the commit hash). Perhaps if you write the intended commit message here I can amend it.

mmcky commented:
oh that definitely sounds like something I'm not going to try :-). 

The initial PR message has more details but here is a good summary

---

✨ NEW: Introducing jupyterbook-latex and added handling of tableofcontents for latex (PR #1167)

This PR brings improvements to `PDF` builds via `LaTeX` by incorporating the [jupyterbook-latex](https://github.com/executablebooks/jupyterbook-latex) extension. This initial release:
1. harmonises the document structure for `html` and `latex` outputs, by implementing support for `part`/`chapter` structures in the `_toc.yml` for PDF via LaTeX, 
2. brings support for the jupyter-book directive `tableofcontents` to PDF via LaTeX, and 
3. adds support for commonly used tags on `code-cell` directives (such as `hide-cell`) to PDF via LaTeX.

chrisjsewell commented:
ok sorted, yeh you shouldn't normally rebase master, because it can screw up other peoples branches/PRs of it. But as this was only just merged, then hopefully no one would have updated their branches with it anyway

AakashGfude commented:
> > @AakashGfude can you organise this as a priority.
> 
> let me know if you need any help, its pretty straightforward

I have to send a PR of the recipe to https://github.com/conda-forge/staged-recipes . am I right?

chrisjsewell commented:
Yep 👍

choldgraf commented:
congrats @AakashGfude on getting this one in 🚀🚀🚀
